xinference/model/llm/pytorch/omnilmm.py

```diff
@@ -0,0 +1,168 @@
+    ChatCompletion,
+    ChatCompletionChoice,
+    ChatCompletionChunk,
+    ChatCompletionMessage,
+    CompletionUsage,
+)
+from ..llm_family import LLMFamilyV1, LLMSpecV1
+from .core import PytorchChatModel, PytorchGenerateConfig
+
+logger = logging.getLogger(__name__)
+
+
+class OmniLMMModel(PytorchChatModel):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self._tokenizer = None
+        self._model = None
+
+    @classmethod
+    def match(
+        cls, model_family: "LLMFamilyV1", model_spec: "LLMSpecV1", quantization: str
+    ) -> bool:
+        if "OmniLMM" in model_family.model_name:
+            return True
+        return False
+
+    def load(self):
+        from transformers import AutoModel, AutoTokenizer
+
+        device = self._pytorch_model_config.get("device", "auto")
+        device = select_device(device)
+
+        self._tokenizer = AutoTokenizer.from_pretrained(
+            self.model_path,
+            trust_remote_code=True,
+            code_revision=self.model_spec.model_revision,
+        )
+
+        self._model = AutoModel.from_pretrained(
+            self.model_path,
+            device_map=device,
+            trust_remote_code=True,
+            code_revision=self.model_spec.model_revision,
+        ).eval()
+
+    def _message_content_to_OmniLMM(
+        self, content
+    ) -> tuple[list[dict[str, str]], list[dict[str, str]]]:
+        def _ensure_url(_url):
+            if _url.startswith("data:"):
+                logging.info("Parse url by base64 decoder.")
+                # https://platform.openai.com/docs/guides/vision/uploading-base-64-encoded-images
+                # e.g. f"data:image/jpeg;base64,{base64_image}"
+                _type, data = _url.split(";")
+                _, ext = _type.split("/")
+                data = data[len("base64,") :]
+                data = base64.b64decode(data.encode("utf-8"))
+
+                with tempfile.NamedTemporaryFile(suffix=f".{ext}", delete=False) as f:
+                    f.write(data)
+                logging.info("Dump base64 data to %s", f.name)
+                return f.name
+            else:
+                if len(_url) > 2048:
+                    raise Exception(f"Image url is too long, {len(_url)} > 2048.")
+                return _url
+
+        if not isinstance(content, str):
+            images = []
+            other_content = []
+
+            for c in content:
+                if c.get("type") == "image_url":
+                    images.append(
+                        {"image": _ensure_url(c["image_url"]["url"]), "type": "image"}
+                    )
+                else:
+                    other_content.append(c)
+
+            images = sorted(images, key=operator.itemgetter("type"))
+            other_content = sorted(other_content, key=operator.itemgetter("type"))
+
+            return images, other_content
+        return [], [{"type": "text", "text": content}]
+
+    def chat(
```

Comment:
Refer to the offical way to impl `chat`:
<img width="1141" alt="image" src="https://github.com/xorbitsai/inference/assets/109656400/ad9b55d9-0441-4e5a-a3f5-87fdf24ab6e7">
You may not use transformer directly, since OmniLMM preprocesses something via its own way.